### ISSUE TITLE
Performance regression from 1.59 to 1.60

### ISSUE THREAD

**Issue by chuigda** [rust-lang/rust#95877]
### Code

I upgraded to rustc 1.60 today and benchmarked my [personal project](https://github.com/Pr47/Pr47). The benchmark
```bash
make bench_al31fm2_ffi
make bench_al31fm2_fib35
```
shown that the compilation product was 10%~15% slower on my machine.

I don't have a minimal reproducible example yet, so I'll collect as many information as I can.
*I have a somewhat minified version of `al31f`, I'll experiment it out as soon as I have spare time.*

### Version it worked on

1.52.1(I guess) ~ 1.59

It most recently worked on: 1.59

### Version with regression

1.60 stable, and most recent nightly

`rustc --version --verbose`:
```
rustc 1.60.0 (7737e0b5c 2022-04-04)
binary: rustc
commit-hash: 7737e0b5c4103216d6fd8cf941b7ab9bdbaace7c
commit-date: 2022-04-04
host: x86_64-unknown-linux-gnu
release: 1.60.0
LLVM version: 14.0.0
```

**Comment by chuigda** [comment]
My CPU info:
```
processor	: 0
vendor_id	: AuthenticAMD
cpu family	: 23
model		: 113
model name	: AMD Ryzen 7 3700X 8-Core Processor
stepping	: 0
microcode	: 0x8701021
cpu MHz		: 3600.000
cache size	: 512 KB
physical id	: 0
siblings	: 16
core id		: 0
cpu cores	: 8
apicid		: 0
initial apicid	: 0
fpu		: yes
fpu_exception	: yes
cpuid level	: 16
wp		: yes
flags		: fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx mmxext fxsr_opt pdpe1gb rdtscp lm constant_tsc rep_good nopl nonstop_tsc cpuid extd_apicid aperfmperf rapl pni pclmulqdq monitor ssse3 fma cx16 sse4_1 sse4_2 movbe popcnt aes xsave avx f16c rdrand lahf_lm cmp_legacy svm extapic cr8_legacy abm sse4a misalignsse 3dnowprefetch osvw ibs skinit wdt tce topoext perfctr_core perfctr_nb bpext perfctr_llc mwaitx cpb cat_l3 cdp_l3 hw_pstate ssbd mba ibpb stibp vmmcall fsgsbase bmi1 avx2 smep bmi2 cqm rdt_a rdseed adx smap clflushopt clwb sha_ni xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local clzero irperf xsaveerptr rdpru wbnoinvd arat npt lbrv svm_lock nrip_save tsc_scale vmcb_clean flushbyasid decodeassists pausefilter pfthreshold avic v_vmsave_vmload vgif v_spec_ctrl umip rdpid overflow_recov succor smca sme sev sev_es
bugs		: sysret_ss_attrs spectre_v1 spectre_v2 spec_store_bypass
bogomips	: 7202.03
TLB size	: 3072 4K pages
clflush size	: 64
cache_alignment	: 64
address sizes	: 43 bits physical, 48 bits virtual
power management: ts ttp tm hwpstate cpb eff_freq_ro [13] [14]
```

Kernel info:
```
Linux iceyself 5.15.28-1-MANJARO #1 SMP PREEMPT Fri Mar 11 14:12:57 UTC 2022 x86_64 GNU/Linux
```

**Comment by chuigda** [comment]
Flamegraphs, raw perf data, and assemblies (created using objdump)

[rustc-1.60-regression.zip](https://github.com/rust-lang/rust/files/8458553/rustc-1.60-regression.zip)

**Comment by apiraino** [comment]
labeling this as "needs an MCVE", a minimal reproducible example would certainly help

**Comment by mwlon** [comment]
I'm also seeing a performance regression, about 35% for one of my workloads, switching from 1.59 to 1.60.

**Comment by mwlon** [comment]
After some investigation, I found 1.60 is doing less inlining on my project. I brought the 35% performance hit down to 4% with a couple of `#[inline(always)]`.

**Comment by chuigda** [comment]
Sorry for long--- time no response. I'm picking my project up and it seems that this regression have been resolved since Rust 1.66. I'm closing.
